With UI-side compositing, many tests hit ASSERT(m_scrollingTrees.contains(webPage->identifier()));

smfr opened the pull request:
#### 82e5917868fa066744774abe4e292959f375fd02
<pre>
With UI-side compositing, many tests hit ASSERT(m_scrollingTrees.contains(webPage-&gt;identifier()));
<a href="https://bugs.webkit.org/show_bug.cgi?id=247766">https://bugs.webkit.org/show_bug.cgi?id=247766</a>
rdar://102214322

Reviewed by Cameron McCormack.

With UI-side compositing, the web process ScrollingCoordinator doesn&apos;t have a ScrollingTree,
so the assertion in `EventDispatcher::removeScrollingTreeForPage()` is wrong.

To fix this, delegate the registration and unregistration of the scrolling tree
to the DrawingArea. Then DrawingArea subclasses that have a WebContent process scrolling
tree can do the registration/unregistration (currently TiledCoreAnimationDrawingArea and
DrawingAreaCoordinatedGraphics).

Make the WebPage argument to the EventDispatcher functions be a reference.

* Source/WebKit/WebProcess/WebPage/DrawingArea.h:
(WebKit::DrawingArea::registerScrollingTree):
(WebKit::DrawingArea::unregisterScrollingTree):
* Source/WebKit/WebProcess/WebPage/EventDispatcher.cpp:
(WebKit::EventDispatcher::addScrollingTreeForPage):
(WebKit::EventDispatcher::removeScrollingTreeForPage):
* Source/WebKit/WebProcess/WebPage/EventDispatcher.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_appHighlightsVisible):
(WebKit::WebPage::close):
* Source/WebKit/WebProcess/WebPage/mac/TiledCoreAnimationDrawingArea.h:
* Source/WebKit/WebProcess/WebPage/mac/TiledCoreAnimationDrawingArea.mm:
(WebKit::TiledCoreAnimationDrawingArea::registerScrollingTree):
(WebKit::TiledCoreAnimationDrawingArea::unregisterScrollingTree):

Canonical link: <a href="https://commits.webkit.org/256571@main">https://commits.webkit.org/256571@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4702c9a0d9a9e66a14947ed29dd74fc23d56963a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/96156 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/5406 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/29207 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/105701 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/5530 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/34172 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/88535 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/101525 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/101815 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/4101 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/82759 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/31127 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/85949 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/87855 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/73957 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/39896 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/19376 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/37570 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/20729 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4552 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/42166 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/43332 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/44057 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/39991 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

heycam commented on Source/WebKit/WebProcess/WebPage/EventDispatcher.cpp:
```diff
@@ -87,7 +87,6 @@ void EventDispatcher::addScrollingTreeForPage(WebPage* webPage)
 void EventDispatcher::removeScrollingTreeForPage(WebPage* webPage)
 {
     Locker locker { m_scrollingTreesLock };
-    ASSERT(m_scrollingTrees.contains(webPage->identifier()));
```
When UI side compositing is in use, if this assertion isn't valid, can we avoid calling removeScrollingTreeForPage entirely? And avoid touching m_scrollingTreesLock.

smfr committed:
With UI-side compositing, many tests hit ASSERT(m_scrollingTrees.contains(webPage->identifier()));
https://bugs.webkit.org/show_bug.cgi?id=247766
rdar://102214322

Reviewed by Cameron McCormack.

With UI-side compositing, the web process ScrollingCoordinator doesn't have a ScrollingTree,
so the assertion in `EventDispatcher::removeScrollingTreeForPage()` is wrong.

To fix this, delegate the registration and unregistration of the scrolling tree
to the DrawingArea. Then DrawingArea subclasses that have a WebContent process scrolling
tree can do the registration/unregistration (currently TiledCoreAnimationDrawingArea and
DrawingAreaCoordinatedGraphics).

Make the WebPage argument to the EventDispatcher functions be a reference.

* Source/WebKit/WebProcess/WebPage/DrawingArea.h:
(WebKit::DrawingArea::registerScrollingTree):
(WebKit::DrawingArea::unregisterScrollingTree):
* Source/WebKit/WebProcess/WebPage/EventDispatcher.cpp:
(WebKit::EventDispatcher::addScrollingTreeForPage):
(WebKit::EventDispatcher::removeScrollingTreeForPage):
* Source/WebKit/WebProcess/WebPage/EventDispatcher.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_appHighlightsVisible):
(WebKit::WebPage::close):
* Source/WebKit/WebProcess/WebPage/mac/TiledCoreAnimationDrawingArea.h:
* Source/WebKit/WebProcess/WebPage/mac/TiledCoreAnimationDrawingArea.mm:
(WebKit::TiledCoreAnimationDrawingArea::registerScrollingTree):
(WebKit::TiledCoreAnimationDrawingArea::unregisterScrollingTree):

Canonical link: https://commits.webkit.org/256571@main
